sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/ThrottleWithExternalResource.java

```diff
@@ -0,0 +1,409 @@
+   * reading and writing to a shared instance. See <a
+   * href="https://redis.io/docs/get-started/faq/">Redis FAQ</a> for more information on important
+   * considerations when using Redis as {@link ThrottleWithExternalResource}'s external resource.
+   */
+  static <@NonNull T>
+      ThrottleWithExternalResource<
+              @NonNull T,
+              RedisReporter,
+              RedisEnqueuer<@NonNull T>,
+              RedisDequeuer<@NonNull T>,
+              RedisDecrementer,
+              RedisRefresher>
+          usingRedis(
+              URI uri,
+              String quotaIdentifier,
+              String queueKey,
+              Quota quota,
+              Coder<@NonNull T> coder)
+              throws Coder.NonDeterministicException {
+    return new ThrottleWithExternalResource<
+        @NonNull T,
+        RedisReporter,
+        RedisEnqueuer<@NonNull T>,
+        RedisDequeuer<@NonNull T>,
+        RedisDecrementer,
+        RedisRefresher>(
+        quota,
+        quotaIdentifier,
+        coder,
+        new RedisReporter(uri),
+        new RedisEnqueuer<>(uri, queueKey, coder),
+        new RedisDequeuer<>(uri, coder, queueKey),
+        new RedisDecrementer(uri, queueKey),
+        new RedisRefresher(uri, quota, quotaIdentifier));
+  }
+
+  private static final Duration THROTTLE_INTERVAL = Duration.standardSeconds(1L);
+
+  private final @NonNull Quota quota;
+  private final @NonNull String quotaIdentifier;
+  private final @NonNull Coder<@NonNull T> coder;
+  private final @NonNull ReporterT reporterT;
+  private final @NonNull EnqueuerT enqueuerT;
+  private final @NonNull DequeuerT dequeuerT;
+  private final @NonNull DecrementerT decrementerT;
+  private final @NonNull RefresherT refresherT;
+
+  ThrottleWithExternalResource(
+      @NonNull Quota quota,
+      @NonNull String quotaIdentifier,
+      @NonNull Coder<@NonNull T> coder,
+      @NonNull ReporterT reporterT,
+      @NonNull EnqueuerT enqueuerT,
+      @NonNull DequeuerT dequeuerT,
+      @NonNull DecrementerT decrementerT,
+      @NonNull RefresherT refresherT)
+      throws Coder.NonDeterministicException {
+    this.quotaIdentifier = quotaIdentifier;
+    this.reporterT = reporterT;
+    coder.verifyDeterministic();
+    checkArgument(!quotaIdentifier.isEmpty());
+    this.quota = quota;
+    this.coder = coder;
+    this.enqueuerT = enqueuerT;
+    this.dequeuerT = dequeuerT;
+    this.decrementerT = decrementerT;
+    this.refresherT = refresherT;
+  }
+
+  @Override
+  public Call.@NonNull Result<@NonNull T> expand(PCollection<@NonNull T> input) {
+    Pipeline pipeline = input.getPipeline();
+
+    // Refresh known quota to control the throttle rate.
+    Call.Result<Void> refreshResult =
+        pipeline
+            .apply("quota/impulse", PeriodicImpulse.create().withInterval(quota.getInterval()))
+            .apply("quota/refresh", getRefresher());
```

Comment:
There is a side effect of using slash to name transform. Slash is used to indicate subtransform hierarchy. On Dataflow UI the jsubtransform is packed in a by same prefix. This will cause some minor glitch, see internal tracker b/295936006